packages/builder/src/dataBinding.js

```diff
@@ -286,44 +282,39 @@ export const getComponentContexts = (
 }
 
 /**
- * Gets all data provider components above a component.
+ * Gets all components available to this component that expose a certain action
  */
 export const getActionProviders = (
   asset,
   componentId,
   actionType,
   options = { includeSelf: false }
 ) => {
-  if (!asset) {
-    return []
-  }
-
-  // Get all components
-  const components = findAllComponents(asset.props)
+  const contexts = getAllComponentContexts(asset, componentId, "action", {
+    includeSelf: options?.includeSelf,
+  })
+  return (
+    contexts
+      // Find the definition of the action in question, if one is provided
+      .map(context => ({
+        ...context,
 
-  // Find matching contexts and generate bindings
-  let providers = []
-  components.forEach(component => {
-    if (!options?.includeSelf && component._id === componentId) {
-      return
-    }
-    const def = componentStore.getDefinition(component._component)
-    const actions = (def?.actions || []).map(action => {
-      return typeof action === "string" ? { type: action } : action
-    })
-    const action = actions.find(x => x.type === actionType)
-    if (action) {
-      let runtimeBinding = component._id
-      if (action.suffix) {
-        runtimeBinding += `-${action.suffix}`
-      }
-      providers.push({
-        readableBinding: component._instanceName,
-        runtimeBinding,
+        action: context.contexts[0].actions.find(x => x.type === actionType),
```

Comment:
NAB: How do you know that you can always use the first context?

Reply:
Hmm, I'll take your word for it 😅 
If they're was a reasonable way to pass a single object instead of a list it might be worth doing that, but no worries if it's a hassle!

Reply:
I've added in some optional chaining to it wouldn't error even if this structure didn't exist. It's 100% safe without this at the moment, but it's sensible to be more cautious since it could potentially change in future and introduce a bug, so this will prevent that happening 👍